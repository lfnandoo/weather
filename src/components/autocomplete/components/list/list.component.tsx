import { Skeleton } from '../../../skeleton/skeleton.component';
import { ListContainer, UlList } from './list.styles';

function List({
  items,
  isLoading,
  onClick,
}: {
  items: { description: string }[];
  isLoading: boolean;
  onClick: (index: number) => void;
}) {
  return (
    <ListContainer>
      <UlList data-testid="suggestions">
        {!isLoading && (
          <>
            {items.map((item, index) => (
              <li data-testid="item" key={Math.random()} onClick={() => onClick(index)}>
                {item.description}
              </li>
            ))}
            {items.length === 0 && <p>Nenhum resultado encontrado</p>}
          </>
        )}
        {isLoading &&
          Array.from({ length: 3 }).map(() => (
            <li data-testid="item-loading" key={Math.random()}>
              <Skeleton width="150px" height="20px" />
            </li>
          ))}
      </UlList>
    </ListContainer>
  );
}

export { List };
