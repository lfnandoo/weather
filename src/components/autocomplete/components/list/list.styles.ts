import styled from 'styled-components';

const ListContainer = styled.div`
  display: grid;
  place-items: center;
  position: absolute;
  height: 80px;
  z-index: 1;
  overflow-y: auto;
  width: 100%;
  padding: 5px 0;
  background-color: #fff;
  border-radius: 5px;
`;

const UlList = styled.ul`
  display: flex;
  align-items: center;
  flex-direction: column;
  gap: 5px;
  list-style: none;
  padding: 5px 0;
  margin: 0 auto;
  background-color: #fff;
  border-radius: 0 0 5px 5px;
  width: 100%;
  position: absolute;
  z-index: 1;
  overflow-y: auto;
`;

export { ListContainer, UlList };
