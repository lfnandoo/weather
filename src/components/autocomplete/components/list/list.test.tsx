import { render } from '@testing-library/react';
import { fireEvent, getByTestId as gGetByTestId } from '@testing-library/dom';
import { List } from './list.component';

describe('Autocomplete > List Component', () => {
  it('should render items', () => {
    const onClick = jest.fn();
    const { getAllByTestId } = render(
      <List
        items={[{ description: 'Teste' }, { description: 'Teste' }, { description: 'Teste' }]}
        onClick={onClick}
        isLoading={false}
      />,
    );

    expect(getAllByTestId('item')).toHaveLength(3);
  });

  it('should render empty state', () => {
    const onClick = jest.fn();
    const { getByTestId } = render(<List items={[]} onClick={onClick} isLoading={false} />);

    expect(getByTestId('empty')).toBeInTheDocument();
  });

  it('should render 3 loading items when is loading', () => {
    const onClick = jest.fn();
    const { queryAllByTestId } = render(<List items={[]} onClick={onClick} isLoading />);

    expect(queryAllByTestId('item-loading')).toHaveLength(3);
  });

  it('should render a single skeleton inside items when is loading', () => {
    const onClick = jest.fn();
    const { getAllByTestId } = render(<List items={[]} onClick={onClick} isLoading />);

    getAllByTestId('item-loading').forEach((element) => {
      expect(element.childElementCount).toBe(1);
      expect(gGetByTestId(element, 'skeleton')).toBeInTheDocument();
    });
  });

  it('should trigger onClick callback when click on a item', () => {
    const onClick = jest.fn();
    const { getByText } = render(
      <List
        items={[{ description: 'Teste' }, { description: 'Teste1' }]}
        onClick={onClick}
        isLoading={false}
      />,
    );

    fireEvent.click(getByText('Teste1'));

    expect(onClick).toHaveBeenCalledWith(1);
  });
});
