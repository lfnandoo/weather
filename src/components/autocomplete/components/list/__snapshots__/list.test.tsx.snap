// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete > List Component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsnTZi dxpncY"
  >
    <ul
      class="sc-dkzDqf ioOdXm"
      data-testid="suggestions"
    >
      <li
        data-testid="item"
      >
        Teste
      </li>
      <li
        data-testid="item"
      >
        Teste
      </li>
      <li
        data-testid="item"
      >
        Teste
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Autocomplete > List Component should render correctly 2`] = `
<DocumentFragment>
  <div
    class="sc-gsnTZi dxpncY"
  >
    <ul
      class="sc-dkzDqf ioOdXm"
      data-testid="suggestions"
    >
      <li
        data-testid="item-loading"
      >
        <div
          class="sc-bczRLJ bEgVHR"
          data-testid="skeleton"
          height="20px"
          width="150px"
        />
      </li>
      <li
        data-testid="item-loading"
      >
        <div
          class="sc-bczRLJ bEgVHR"
          data-testid="skeleton"
          height="20px"
          width="150px"
        />
      </li>
      <li
        data-testid="item-loading"
      >
        <div
          class="sc-bczRLJ bEgVHR"
          data-testid="skeleton"
          height="20px"
          width="150px"
        />
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
